lib/ui/volatile_path_tracker.cc

```diff
@@ -0,0 +1,74 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "flutter/lib/ui/volatile_path_tracker.h"
+
+namespace flutter {
+
+VolatilePathTracker::VolatilePathTracker(
+    fml::RefPtr<fml::TaskRunner> ui_task_runner)
+    : ui_task_runner_(ui_task_runner) {}
+
+void VolatilePathTracker::Insert(std::shared_ptr<Path> path) {
+  FML_DCHECK(ui_task_runner_->RunsTasksOnCurrentThread());
+  FML_DCHECK(path);
+  FML_DCHECK(path->path.isVolatile());
+  paths_.insert(path);
+}
+
+void VolatilePathTracker::Erase(std::shared_ptr<Path> path) {
+  FML_DCHECK(path);
+  if (ui_task_runner_->RunsTasksOnCurrentThread()) {
+    paths_.erase(path);
+    return;
+  }
+
+  std::scoped_lock lock(paths_to_remove_mutex_);
+  needs_drain_ = true;
+  paths_to_remove_.push_back(path);
+}
+
+void VolatilePathTracker::OnFrame() {
+  FML_DCHECK(ui_task_runner_->RunsTasksOnCurrentThread());
+  std::string total_count = std::to_string(paths_.size());
+  TRACE_EVENT1("flutter", "VolatilePathTracker::OnFrame", "count",
+               total_count.c_str());
+
+  Drain();
+
+  for (auto it = paths_.begin(), last = paths_.end(); it != last;) {
```

Comment:
`unordered_set::erase` ensures that iterators (other than the erased iterator) remain valid and the iteration order is unchanged (see https://en.cppreference.com/w/cpp/container/unordered_set/erase)

`unordered_set` iterators are const, so the elements in the set can not be mutated during iteration.  It would be clearer if `path` is explictly declared as a `const std::shared_ptr<Path>` instead of using `auto`.

Reply:
> unordered_set::erase ensures that iterators (other than the erased iterator) remain valid and the iteration order is unchanged (see https://en.cppreference.com/w/cpp/container/unordered_set/erase)

I read that as the validity of existing iterators. But what about the order of iteration? The result of the `++` I mean. Perhaps I am confusing this. But just using a separate set with a swap should be faster and easier to reason about right?

Reply:
I'll go ahead and refactor to use a set for now.

Reply:
Jason is correct. Quoting from the linked documentation: "The order of the elements that are not erased is preserved. (This makes it possible to erase individual elements while iterating through the container.)"